src/Kestrel.Core/Internal/Http/Http1Frame.cs

```diff
@@ -0,0 +1,447 @@
+            }
+        }
+
+        public void OnHeader(Span<byte> name, Span<byte> value)
+        {
+            _requestHeadersParsed++;
+            if (_requestHeadersParsed > ServerOptions.Limits.MaxRequestHeaderCount)
+            {
+                ThrowRequestRejected(RequestRejectionReason.TooManyHeaders);
+            }
+            var valueString = value.GetAsciiStringNonNullCharacters();
+
+            FrameRequestHeaders.Append(name, valueString);
+        }
+
+        protected void EnsureHostHeaderExists()
+        {
+            if (_httpVersion == Http.HttpVersion.Http10)
+            {
+                return;
+            }
+
+            // https://tools.ietf.org/html/rfc7230#section-5.4
+            // A server MUST respond with a 400 (Bad Request) status code to any
+            // HTTP/1.1 request message that lacks a Host header field and to any
+            // request message that contains more than one Host header field or a
+            // Host header field with an invalid field-value.
+
+            var host = FrameRequestHeaders.HeaderHost;
+            if (host.Count <= 0)
+            {
+                ThrowRequestRejected(RequestRejectionReason.MissingHostHeader);
+            }
+            else if (host.Count > 1)
+            {
+                ThrowRequestRejected(RequestRejectionReason.MultipleHostHeaders);
+            }
+            else if (_requestTargetForm == HttpRequestTarget.AuthorityForm)
+            {
+                if (!host.Equals(RawTarget))
+                {
+                    ThrowRequestRejected(RequestRejectionReason.InvalidHostHeader, host.ToString());
+                }
+            }
+            else if (_requestTargetForm == HttpRequestTarget.AbsoluteForm)
+            {
+                // If the target URI includes an authority component, then a
+                // client MUST send a field - value for Host that is identical to that
+                // authority component, excluding any userinfo subcomponent and its "@"
+                // delimiter.
+
+                // System.Uri doesn't not tell us if the port was in the original string or not.
+                // When IsDefaultPort = true, we will allow Host: with or without the default port
+                var authorityAndPort = _absoluteRequestTarget.Authority + ":" + _absoluteRequestTarget.Port;
+                if ((host != _absoluteRequestTarget.Authority || !_absoluteRequestTarget.IsDefaultPort)
+                    && host != authorityAndPort)
+                {
+                    ThrowRequestRejected(RequestRejectionReason.InvalidHostHeader, host.ToString());
+                }
+            }
+        }
+
+        protected override string CreateRequestId()
+            => StringUtilities.ConcatAsHexSuffix(ConnectionId, ':', _requestCount);
+
+        protected override void OnReset()
+        {
+            ExtraFeatureSet(typeof(IHttpUpgradeFeature), this);
```

Comment:
Is this why you don't like the current design? I definitely don't like allocating the MaybeExtra by default for each Frame object.

Reply:
Actually I hadn't realized this would allocate every time, but I felt the multiple `.FeatureCollection.cs` files kinda smelled. Better have it all in `Frame.FeatureCollection.cs` and null out features that don't belong to a protocol version, right?

Reply:
I don't see the problem with multiple FeatureCollection.cs files.